flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/cache/CachedDataInputStream.java

```diff
@@ -102,62 +102,81 @@ private FSDataInputStream getStream() throws IOException {
         if (isFlinkThread()) {
             cacheEntry.touch();
         }
-        FSDataInputStream stream = tryGetCacheStream();
-        if (stream != null) {
-            fileBasedCache.incHitCounter();
-            return stream;
-        }
-
-        if (streamStatus == StreamStatus.CACHED_CLOSED
-                || streamStatus == StreamStatus.CACHED_CLOSING) {
-            if (streamStatus == StreamStatus.CACHED_CLOSING) {
-                try {
-                    semaphore.acquire(1);
-                } catch (InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-                originalStream.seek(position);
-                position = -1;
-                LOG.trace(
-                        "Stream {} status from {} to {}",
-                        cacheEntry.cachePath,
-                        streamStatus,
-                        StreamStatus.CACHED_CLOSED);
-                streamStatus = StreamStatus.CACHED_CLOSED;
-            }
-            // try reopen
-            tryReopen();
-            stream = tryGetCacheStream();
+        // Repeat until get a stream
+        while (true) {
+            // Firstly, we try to get cache stream
+            FSDataInputStream stream = tryGetCacheStream();
             if (stream != null) {
                 fileBasedCache.incHitCounter();
                 return stream;
             }
-            fileBasedCache.incMissCounter();
-            return originalStream;
-        } else if (streamStatus == StreamStatus.ORIGINAL) {
-            fileBasedCache.incMissCounter();
-            return originalStream;
-        } else {
-            if (streamStatus == StreamStatus.CACHED_OPEN) {
-                stream = tryGetCacheStream();
+
+            // No cache stream, so is it closing?
+            if (streamStatus == StreamStatus.CACHED_CLOSED
+                    || streamStatus == StreamStatus.CACHED_CLOSING) {
+                if (streamStatus == StreamStatus.CACHED_CLOSING) {
+                    // if closing, update the position
+                    try {
+                        semaphore.acquire(1);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                    originalStream.seek(position);
+                    position = -1;
+                    LOG.trace(
+                            "Stream {} status from {} to {}",
+                            cacheEntry.cachePath,
+                            streamStatus,
+                            StreamStatus.CACHED_CLOSED);
+                    streamStatus = StreamStatus.CACHED_CLOSED;
+                }
+                // try reopen
+                stream = tryReopen();
                 if (stream != null) {
                     fileBasedCache.incHitCounter();
                     return stream;
                 }
+                fileBasedCache.incMissCounter();
```

Comment:
Add `streamStatus = StreamStatus.ORIGINAL` here?

Reply:
I suggest not. The `ORIGINAL` means this stream could only be read via original stream, no cache mechanism enabled. It's a stablized status without transition. There is an `else if` branch under this to handle the `ORIGINAL` condition

Reply:
👌 thanks for the clarification.